### PR TITLE
Tests for GreenPlayer

### PR DESCRIPTION
**Pull request by gnarf**

### CODE REVIEW

File: src/GreenPlayer.js [hunk 1]
```diff
@@ -27,12 +27,15 @@ class SoundPlayer extends EventEmitter {
...
         this.outputNode = null;
+        this.volumeEffect = null;
         this.target = null;
 
         this.initialized = false;
         this.isPlaying = false;
         this.isStarting = null;
         this.playbackRate = 1;
+
+        this.handleEvent = this.handleEvent.bind(this);
```
Comment (mzgoddard) [inline src/GreenPlayer.js]: We don't have other comments here >_<. But this could definitely use one.

Comment (mzgoddard) [reply src/GreenPlayer.js]: Like that handleEvent follows the event object style handler but the testing mock doesn't support object style.

File: src/GreenPlayer.js [hunk 2]
```diff
@@ -104,7 +105,12 @@ class SoundPlayer extends EventEmitter {
             return;
         }
 
-        this.volumeEffect.connect(target);
+        if (this.volumeEffect) {
```
Comment (mzgoddard) [inline src/GreenPlayer.js]: nit: explicit test if it isn't a boolean being tested?

File: src/GreenPlayer.js [hunk 3]
```diff
@@ -119,8 +125,10 @@ class SoundPlayer extends EventEmitter {
 
         this.stopImmediately();
 
-        this.volumeEffect.dispose();
-        this.volumeEffect = null;
+        if (this.volumeEffect) {
```
Comment (mzgoddard) [inline src/GreenPlayer.js]: nit: explicit test if it isn't a boolean being tested?

File: src/GreenPlayer.js [hunk 4]
```diff
@@ -223,13 +224,16 @@ class SoundPlayer extends EventEmitter {
...
         }
 
-        this.volumeEffect.set(0);
-        this.outputNode.stop(this.audioEngine.audioContext.currentTime + this.audioEngine.DECAY_TIME);
+        // always do a manual stop on a taken / volume effect fade out sound player
+        // take will emit "stop" as well as reset all of our playing statuses / remove our
+        // nodes / etc
+        const taken = this.take();
+        taken.volumeEffect = new VolumeEffect(taken.audioEngine, taken, null);
+        taken.volumeEffect.connect(taken.target);
```
Comment (mzgoddard) [inline src/GreenPlayer.js]: Let this be implicitly handled by the following `connect()`?

File: src/GreenPlayer.js [hunk 5]
```diff
@@ -141,25 +149,22 @@ class SoundPlayer extends EventEmitter {
...
             taken.isStarting = this.isStarting;
             taken.isPlaying = this.isPlaying;
-            taken.initialize();
-            taken.outputNode.disconnect();
+            taken.initialized = this.initialized;
             taken.outputNode = this.outputNode;
-            taken.outputNode.addEventListener(ON_ENDED, taken);
-            taken.volumeEffect.set(this.volumeEffect.value);
+            taken.outputNode.addEventListener(ON_ENDED, taken.handleEvent);
+            taken.volumeEffect = this.volumeEffect;
```
Comment (mzgoddard) [inline src/GreenPlayer.js:174]: My first feeling on this was to change it back. But rather I like this.

File: src/GreenPlayer.js [hunk 6]
```diff
@@ -188,19 +193,15 @@ class SoundPlayer extends EventEmitter {
...
-            // short period until its volume is 0 and release it to be
-            // eventually garbage collected.
-            this.take().stop();
+            this.stop();
         }
 
-        if (!this.initialized) {
-            this.initialize();
-        } else {
+        if (this.initialized) {
```
Comment (mzgoddard) [inline src/GreenPlayer.js:212]: :heart: the test inversion.

**Review by mzgoddard — APPROVED** [review]
:heart:

**Comment by gnarf** [comment]
Kind of waiting for #89 to land to rebase this on it, if we hit merge, it will join up with #89 though :)

File: src/AudioEngine.js [hunk 7]
```diff
@@ -65,6 +65,10 @@ class AudioEngine {
         this.loudness = null;
     }
 
+    get currentTime () {
```
Comment (gnarf) [inline src/AudioEngine.js]: Add docblock

**Pull request merged**